src/crypto/ocb_internal.cc

```diff
@@ -358,30 +362,82 @@
 #if USE_OPENSSL_AES
 /*---------------*/
 
-#include <openssl/aes.h>                            /* http://openssl.org/ */
+#include <openssl/evp.h>                            /* http://openssl.org/ */
 
 namespace ocb_aes {
 
-typedef AES_KEY KEY;
+typedef EVP_CIPHER_CTX KEY;
 
-static KEY *KEY_new() { return new KEY; }
+enum { BLOCK_SIZE = 16 };
 
-static void KEY_delete(KEY *key) { delete key; }
+static KEY *KEY_new() {
+	KEY *key = EVP_CIPHER_CTX_new();
+	if (key == NULL) {
+		throw std::bad_alloc();
+	}
+	return key;
+}
+
+static void KEY_delete(KEY *key) { EVP_CIPHER_CTX_free(key); }
 
 static void set_encrypt_key(const unsigned char *user_key, int bits, KEY *key) {
-	AES_set_encrypt_key(user_key, bits, key);
+	// This function and the others in this section replicate the behavior of
+	// OpenSSL's deprecated AES_* primitives. Those primitives implemented AES
+	// without any block cipher mode--that is, in ECB mode. Normally, using ECB
+	// mode anywhere would be questionable, but it's safe here because it's
+	// being used to implement a higher-level cryptographic mode (OCB mode),
+	// which is in turn used by Mosh.
+
+	fatal_assert(bits == 128);
+	if (EVP_EncryptInit_ex(key, EVP_aes_128_ecb(), /*impl=*/NULL, user_key, /*iv=*/NULL) != 1 ||
+			EVP_CIPHER_CTX_set_padding(key, false) != 1) {
+		throw Crypto::CryptoException("Could not initialize AES encryption context.");
+	}
 }
 
 static void set_decrypt_key(const unsigned char *user_key, int bits, KEY *key) {
-	AES_set_decrypt_key(user_key, bits, key);
+	// See note in set_encrypt_key about the use of ECB mode.
+	fatal_assert(bits == 128);
+	if (EVP_DecryptInit_ex(key, EVP_aes_128_ecb(), /*impl=*/NULL, user_key, /*iv=*/NULL) != 1 ||
+			EVP_CIPHER_CTX_set_padding(key, false) != 1) {
+		throw Crypto::CryptoException("Could not initialize AES decryption context.");
+	}
 }
 
 static void encrypt(const unsigned char *in, unsigned char *out, KEY *key) {
-	return AES_encrypt(in, out, key);
+	if (EVP_EncryptInit_ex(key, /*type=*/NULL, /*impl=*/NULL, /*key=*/NULL, /*iv=*/NULL) != 1) {
```

Comment:
Add a comment: even though we're in ECB mode, the EVP_CIPHER API requires a call to EncryptInit after every call ot EncryptFinal to reset the context for use in the next block. Since ECB has no IV, and the key is unchanged, this EncryptInit call takes NULL as every parameter.

Reply:
Done.